File: competitiveProgramming/editor/cf.go
package main

import (
    // "bufio"
    "encoding/json"
    "fmt"
    "github.com/atotto/clipboard"
    "github.com/PuerkitoBio/goquery"
    // "io"
    "io/ioutil"
    "net/http"
    "os"
    "os/exec"
    // "path/filepath"
    "regexp"
    "strings"
    "time"
)

const LINKS_FILE = "saved_links.txt"

// --------------------- ADD LINK ---------------------

func addLink(url string) {
    f, err := os.OpenFile(LINKS_FILE, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
    if err != nil {
        fmt.Println("❌ Could not open saved_links.txt:", err)
        return
    }
    defer f.Close()

    if _, err = f.WriteString(url + "\n"); err != nil {
        fmt.Println("❌ Could not write to saved_links.txt:", err)
        return
    }
    fmt.Println("✅ Added link:", url)
}

// --------------------- LOAD PROBLEM ---------------------

func loadProblem(url string) {
    if url == "" {
        fmt.Println("No URL provided.")
        return
    }

    // Last two segments (like 2055/C)
    parts := strings.Split(strings.TrimSpace(url), "/")
    segCount := len(parts)
    if segCount < 2 {
        fmt.Println("URL seems invalid.")
        return
    }
    urlSegment := strings.Join(parts[segCount-2:], "/")

    // Write segment to saved_links.txt
    err := ioutil.WriteFile(LINKS_FILE, []byte(urlSegment+"\n"), 0644)
    if err != nil {
        fmt.Println("❌ Could not write to saved_links.txt:", err)
        return
    }

    // Fetch the page
    client := &http.Client{}
    req, _ := http.NewRequest("GET", url, nil)
    req.Header.Set("User-Agent", "Mozilla/5.0")
    resp, err := client.Do(req)
    if err != nil {
        fmt.Println("❌ HTTP error:", err)
        return
    }
    defer resp.Body.Close()

    if resp.StatusCode != 200 {
        fmt.Printf("❌ Non-200 response: %d\n", resp.StatusCode)
        return
    }

    doc, err := goquery.NewDocumentFromReader(resp.Body)
    if err != nil {
        fmt.Println("❌ goquery parse error:", err)
        return
    }

    // We'll replicate your logic:
    // 1) Find all div.test-example-line-<idx> -> group lines by index
    // 2) Find all "Output" -> next <pre> -> outputs
    reLine := regexp.MustCompile(`test-example-line-(\d+)`)

    inputsMap := make(map[int][]string)

    // find all divs with class ~ test-example-line
    doc.Find("div").Each(func(i int, s *goquery.Selection) {
        class, _ := s.Attr("class")
        if strings.Contains(class, "test-example-line") {
            matches := reLine.FindStringSubmatch(class)
            if len(matches) == 2 {
                idx := matches[1] // string of the number
                // convert to int
                // but let's store them in a map using the int
                // We can parse it:
                // We'll ignore parse error if any
                var iidx int
                fmt.Sscanf(idx, "%d", &iidx)
                text := strings.TrimSpace(s.Text())
                inputsMap[iidx] = append(inputsMap[iidx], text)
            }
        }
    })

    // find all "Output" div.title => next <pre>
    var outputs []string
    doc.Find("div.title").Each(func(i int, s *goquery.Selection) {
        if strings.TrimSpace(s.Text()) == "Output" {
            pre := s.NextFiltered("pre")
            if pre != nil {
                outputs = append(outputs, strings.TrimSpace(pre.Text()))
            }
        }
    })

    // write input.txt
    inFile, err := os.Create("input.txt")
    if err != nil {
        fmt.Println("❌ Could not create input.txt:", err)
        return
    }
    defer inFile.Close()

    var idxList []int
    for k := range inputsMap {
        idxList = append(idxList, k)
    }
    // Sort them
    // we can do a bubble or just create a small function
    // but let's do "sort.Ints"
    importSortInts(idxList)

    for _, idxVal := range idxList {
        lines := inputsMap[idxVal]
        for _, line := range lines {
            inFile.WriteString(line + "\n")
        }
    }

    // write expected.txt
    outFile, err := os.Create("expected.txt")
    if err != nil {
        fmt.Println("❌ Could not create expected.txt:", err)
        return
    }
    defer outFile.Close()

    for _, outLine := range outputs {
        outFile.WriteString(outLine + "\n")
    }

    fmt.Println("✅ Test cases extracted.")
}

// custom int sorting without importing a big package
func importSortInts(a []int) {
    for i := 0; i < len(a); i++ {
        for j := i + 1; j < len(a); j++ {
            if a[i] > a[j] {
                a[i], a[j] = a[j], a[i]
            }
        }
    }
}

// --------------------- COMPILE & RUN ---------------------

func compileCpp() bool {
    cmd := exec.Command("g++", "-std=c++20", "-O2", "-o", "workspace", "workspace.cpp")
    err := cmd.Run()
    return err == nil
}

func runCpp() bool {
    if !compileCpp() {
        fmt.Println("❌ Compilation failed.")
        return false
    }

    if _, err := os.Stat("input.txt"); os.IsNotExist(err) {
        fmt.Println("❌ input.txt not found.")
        return false
    }

    // run ./workspace with input.txt
    inFile, err := os.Open("input.txt")
    if err != nil {
        fmt.Println("❌ Could not open input.txt:", err)
        return false
    }
    defer inFile.Close()

    cmd := exec.Command("./workspace")
    cmd.Stdin = inFile
    outBytes, err := cmd.Output()
    if err != nil {
        fmt.Println("❌ Runtime error:", err)
        return false
    }
    outputLines := strings.Split(string(outBytes), "\n")
    for i := range outputLines {
        outputLines[i] = strings.TrimSpace(outputLines[i])
    }

    if _, err := os.Stat("expected.txt"); os.IsNotExist(err) {
        fmt.Println("❌ expected.txt not found.")
        return false
    }
    expBytes, _ := ioutil.ReadFile("expected.txt")
    expLines := strings.Split(string(expBytes), "\n")

    var filteredExp []string
    for _, l := range expLines {
        line := strings.TrimSpace(l)
        if line != "" {
            filteredExp = append(filteredExp, line)
        }
    }

    fmt.Println("\n📤 Output:")
    for _, line := range outputLines {
        fmt.Println(line)
    }

    fmt.Println("\n✅ Checking output...\n")
    allPassed := true
    maxLen := len(outputLines)
    if len(filteredExp) > maxLen {
        maxLen = len(filteredExp)
    }

    for i := 0; i < maxLen; i++ {
        var outLine, expLine string
        if i < len(outputLines) {
            outLine = outputLines[i]
        }
        if i < len(filteredExp) {
            expLine = filteredExp[i]
        }
        if outLine == expLine {
            fmt.Printf("\033[32m+ Case %d: %s\033[0m\n", i+1, outLine)
        } else {
            fmt.Printf("\033[31m- Case %d: %s\033[0m\n", i+1, outLine)
            fmt.Printf("\033[33m  Expected: %s\033[0m\n", expLine)
            allPassed = false
        }
    }

    if allPassed {
        fmt.Println("\n\033[32m✅ Passed\033[0m")
    } else {
        fmt.Println("\n\033[31m❌ Wrong Answer\033[0m")
    }
    return allPassed
}

// --------------------- CHECK PROBLEM ---------------------

func checkProblem() {
    if runCpp() {
        submitProblem()
    }
}

// --------------------- SUBMIT PROBLEM ---------------------

func submitProblem() {
    if _, err := os.Stat(LINKS_FILE); os.IsNotExist(err) {
        fmt.Println("❌ No saved URLs.")
        return
    }

    data, _ := ioutil.ReadFile(LINKS_FILE)
    urlFrag := strings.TrimSpace(string(data))
    if urlFrag == "" {
        fmt.Println("❌ Empty saved URL.")
        return
    }

    // open with librewolf
    subURL := fmt.Sprintf("https://codeforces.com/problemset/submit/%s", urlFrag)
    exec.Command("librewolf", subURL).Start()

    // copy workspace.cpp
    if _, err := os.Stat("workspace.cpp"); os.IsNotExist(err) {
        fmt.Println("❌ workspace.cpp not found.")
        return
    }
    codeBytes, _ := ioutil.ReadFile("workspace.cpp")
    clipboard.WriteAll(string(codeBytes))
    fmt.Println("✅ Copied workspace.cpp to clipboard.")
}

// --------------------- FRIENDS ---------------------

func friendsCmd() {
    if _, err := os.Stat(LINKS_FILE); os.IsNotExist(err) {
        fmt.Println("❌ No saved URLs.")
        return
    }
    data, _ := ioutil.ReadFile(LINKS_FILE)
    line := strings.TrimSpace(string(data))
    parts := strings.Split(line, "/")
    contestID := parts[0]
    // open link
    frURL := fmt.Sprintf("https://codeforces.com/contest/%s/standings/friends/true", contestID)
    exec.Command("librewolf", frURL).Start()
}

// --------------------- CHECK LAST SUBMISSION ---------------------

func checkLastSubmission() {
    handle := "worthNothing"
    apiURL := fmt.Sprintf("https://codeforces.com/api/user.status?handle=%s&from=1&count=1", handle)

    resp, err := http.Get(apiURL)
    if err != nil {
        fmt.Println("❌ API error:", err)
        return
    }
    defer resp.Body.Close()

    var result struct {
        Status string `json:"status"`
        Comment string `json:"comment"`
        Result []map[string]interface{} `json:"result"`
    }

    dec := json.NewDecoder(resp.Body)
    if err := dec.Decode(&result); err != nil {
        fmt.Println("❌ JSON decode error:", err)
        return
    }

    if result.Status != "OK" || len(result.Result) == 0 {
        fmt.Println("❌ API error or no submissions.")
        return
    }

    sub := result.Result[0]
    prob, _ := sub["problem"].(map[string]interface{})
    contestId, _ := prob["contestId"].(float64)
    index, _ := prob["index"].(string)
    name, _ := prob["name"].(string)
    verdict, _ := sub["verdict"].(string)
    passedTestCount, _ := sub["passedTestCount"].(float64)
    timeConsumed, _ := sub["timeConsumedMillis"].(float64)
    memoryUsed, _ := sub["memoryConsumedBytes"].(float64)

    fmt.Printf("📘 Problem: %.0f%s - %s\n", contestId, index, name)
    if verdict == "" {
        verdict = "N/A"
    }
    fmt.Printf("🧪 Verdict: %s\n", verdict)
    fmt.Printf("✅ Passed: %.0f\n", passedTestCount)
    fmt.Printf("⚡ Time: %.0f ms\n", timeConsumed)
    fmt.Printf("📦 Memory: %.0f bytes\n", memoryUsed)
}

// --------------------- SHOW UPCOMING REGULAR CONTESTS ---------------------

func showUpcomingRegularContests() {
    apiURL := "https://codeforces.com/api/contest.list?gym=false"
    resp, err := http.Get(apiURL)
    if err != nil {
        fmt.Println("❌ HTTP error:", err)
        return
    }
    defer resp.Body.Close()

    var data struct {
        Status string `json:"status"`
        Comment string `json:"comment"`
        Result []map[string]interface{} `json:"result"`
    }
    if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
        fmt.Println("❌ JSON decode error:", err)
        return
    }
    if data.Status != "OK" {
        fmt.Println("❌ Failed:", data.Comment)
        return
    }

    // Filter upcoming
    upcoming := []map[string]interface{}{}
    for _, c := range data.Result {
        if c["phase"] == "BEFORE" {
            upcoming = append(upcoming, c)
        }
    }
    if len(upcoming) == 0 {
        fmt.Println("📭 No upcoming contests.")
        return
    }

    // Sort by startTimeSeconds
    for i := 0; i < len(upcoming); i++ {
        for j := i+1; j < len(upcoming); j++ {
            si := upcoming[i]["startTimeSeconds"].(float64)
            sj := upcoming[j]["startTimeSeconds"].(float64)
            if si > sj {
                upcoming[i], upcoming[j] = upcoming[j], upcoming[i]
            }
        }
    }

    fmt.Println("\n📅 Upcoming Contests (IST):\n")
    for _, c := range upcoming {
        name, _ := c["name"].(string)
        startF, _ := c["startTimeSeconds"].(float64)
        durationF, _ := c["durationSeconds"].(float64)

        // Convert to IST
        startSec := int64(startF)
        durHours := int64(durationF / 3600)
        // define IST offset
        ist := time.FixedZone("IST", 5*3600+30*60)
        startTime := time.Unix(startSec, 0).UTC().In(ist)

        fmt.Printf("📌 %s | 🕒 %s IST | ⏱️ %dh\n",
            name,
            startTime.Format("2006-01-02 15:04"),
            durHours,
        )
    }
}

// --------------------- MAIN ---------------------

func main() {
    if len(os.Args) < 2 {
        fmt.Println("Usage: go run cf.go <command> [args]")
        return
    }
    cmd := os.Args[1]
    var arg string
    if len(os.Args) > 2 {
        arg = os.Args[2]
    }

    switch cmd {
    case "add":
        if arg == "" {
            fmt.Println("❌ Missing URL for 'add'")
            return
        }
        addLink(arg)

    case "load":
        if arg == "" {
            fmt.Println("❌ Missing URL for 'load'")
            return
        }
        loadProblem(arg)

    case "check":
        checkProblem()

    case "submit":
        submitProblem()

    case "friends":
        friendsCmd()

    case "last":
        checkLastSubmission()

    case "contest":
        showUpcomingRegularContests()

    default:
        fmt.Printf("❓ Unknown or incomplete command '%s'\n", cmd)
    }
}

File: competitiveProgramming/editor/cf.py
import os
import re
import subprocess
import sys
from datetime import datetime, timedelta, timezone

import httpx
import pyperclip
from bs4 import BeautifulSoup

LINKS_FILE = "saved_links.txt"


def add_link(url):
    with open(LINKS_FILE, "a", encoding="utf-8") as f:
        f.write(url + "\n")


def load_problem(url):
    if not url:
        print("No URL provided.")
        return

    url_segment = "/".join(url.strip().split("/")[-2:])
    with open(LINKS_FILE, "w", encoding="utf-8") as f:
        f.write(url_segment + "\n")

    r = httpx.get(url, headers={"User-Agent": "Mozilla/5.0"})
    soup = BeautifulSoup(r.text, "html.parser")

    inputs, outputs = {}, []

    for div in soup.find_all("div", class_=re.compile(r"test-example-line")):
        for c in div.get("class", []):
            if m := re.match(r"test-example-line-(\d+)", c):
                idx = int(m.group(1))
                inputs.setdefault(idx, []).append(div.get_text(strip=True))
                break

    for out_div in soup.find_all("div", class_="title", string="Output"):
        pre = out_div.find_next("pre")
        if pre:
            outputs.append(pre.get_text(strip=True))

    with open("input.txt", "w", encoding="utf-8") as f:
        for idx in sorted(inputs):
            f.write("\n".join(inputs[idx]) + "\n")

    with open("expected.txt", "w", encoding="utf-8") as f:
        f.write("\n".join(outputs))

    print("✅ Test cases extracted.")


def compile_cpp():
    return (
        subprocess.run(
            ["g++", "-std=c++20", "-O2", "-o", "workspace", "workspace.cpp"]
        ).returncode
        == 0
    )


def run():
    if not compile_cpp():
        print("❌ Compilation failed.")
        return False

    if not os.path.exists("input.txt"):
        print("❌ input.txt not found.")
        return False

    with open("input.txt") as fin:
        proc = subprocess.run(
            ["./workspace"], stdin=fin, capture_output=True, text=True
        )
    output = [line.strip() for line in proc.stdout.splitlines()]

    if not os.path.exists("expected.txt"):
        print("❌ expected.txt not found.")
        return False

    with open("expected.txt") as f:
        expected = [line.strip() for line in f if line.strip()]

    print("\n📤 Output:")
    for line in output:
        print(line)

    print("\n✅ Checking output...\n")
    all_passed = True
    for i in range(max(len(output), len(expected))):
        out = output[i] if i < len(output) else ""
        exp = expected[i] if i < len(expected) else ""
        if out == exp:
            print(f"\033[32m+ Case {i+1}: {out}\033[0m")
        else:
            print(f"\033[31m- Case {i+1}: {out}\033[0m")
            print(f"\033[33m  Expected: {exp}\033[0m")
            all_passed = False

    print(
        "\n\033[32m✅ Passed\033[0m"
        if all_passed
        else "\n\033[31m❌ Wrong Answer\033[0m"
    )
    return all_passed


def check_problem():
    if run():
        submit_problem()


def submit_problem():
    if not os.path.exists(LINKS_FILE):
        print("❌ No saved URLs.")
        return

    with open(LINKS_FILE) as f:
        url_fragment = f.readline().strip()

    if not url_fragment:
        print("❌ Empty saved URL.")
        return

    subprocess.run(
        ["librewolf", f"https://codeforces.com/problemset/submit/{url_fragment}"]
    )

    if os.path.exists("workspace.cpp"):
        with open("workspace.cpp", encoding="utf-8") as f:
            pyperclip.copy(f.read())
    else:
        print("❌ workspace.cpp not found.")


def friends():
    with open(LINKS_FILE) as f:
        contest_id = f.readline().strip().split("/")[0]
    subprocess.run(
        [
            "librewolf",
            f"https://codeforces.com/contest/{contest_id}/standings/friends/true",
        ]
    )


def check_last_submission():
    handle = "worthNothing"
    url = f"https://codeforces.com/api/user.status?handle={handle}&from=1&count=1"
    res = httpx.get(url).json()

    if res["status"] != "OK" or not res["result"]:
        print("❌ API error.")
        return

    sub = res["result"][0]
    p = sub["problem"]
    print(f"📘 Problem: {p['contestId']}{p['index']} - {p['name']}")
    print(f"🧪 Verdict: {sub.get('verdict', 'N/A')}")
    print(f"✅ Passed: {sub.get('passedTestCount', 'N/A')}")
    print(f"⚡ Time: {sub.get('timeConsumedMillis', 0)} ms")
    print(f"📦 Memory: {sub.get('memoryConsumedBytes', 0)} bytes")


def show_upcoming_regular_contests():
    url = "https://codeforces.com/api/contest.list?gym=false"
    try:
        res = httpx.get(url).json()
        if res["status"] != "OK":
            print("❌ Failed:", res.get("comment", "No message"))
            return

        upcoming = [c for c in res["result"] if c["phase"] == "BEFORE"]
        if not upcoming:
            print("📭 No upcoming contests.")
            return

        print("\n📅 Upcoming Contests (IST):\n")
        for c in sorted(upcoming, key=lambda x: x["startTimeSeconds"]):
            name = c["name"]
            ist = datetime.fromtimestamp(
                c["startTimeSeconds"], timezone.utc
            ) + timedelta(hours=5, minutes=30)
            print(
                f"📌 {name} | 🕒 {ist.strftime('%Y-%m-%d %H:%M')} IST | ⏱️ {c['durationSeconds'] // 3600}h"
            )

    except Exception as e:
        print("❌ Error:", e)


def main():
    if len(sys.argv) < 2:
        print("Usage: python cf.py <command> [link]")
        return

    cmd = sys.argv[1]
    arg = sys.argv[2] if len(sys.argv) > 2 else None

    match cmd:
        case "add" if arg:
            add_link(arg)
        case "load" if arg:
            load_problem(arg)
        case "check":
            check_problem()
        case "submit":
            submit_problem()
        case "contest":
            show_upcoming_regular_contests()
        case "last":
            check_last_submission()
        case "friends":
            friends()
        case _:
            print(f"❓ Unknown or incomplete command '{cmd}'")


if __name__ == "__main__":
    main()
